Only load Stripe JS on product page when required

reykjalin opened the pull request:
## Changes proposed in this Pull Request:

Fixes #2065 

- Don't load Stripe JS on product pages when PRBs are disabled.
- Don't load Stripe JS on the cart page when PRBs are disabled.

## Testing instructions

> **Note:** The test should be performed with UPE enabled and disabled, and on both the Shortcode and Block-based cart pages.

1. Make sure PRBs are only present on the **Checkout** page.
2. `git switch develop`.
3. Open the **Network** tab of your browser's developer tools.
4. Open any product details page.
5. Notice that we load scripts from js.stripe.com.
6. Add a product to your cart and go to your store's cart page.
7. Notice that we load scripts from js.stripe.com.
8. `git switch fix/2065-only-load-stripe-js-on-product-pages-when-required`
9. Open any product details page.
10. Make sure we still load scripts from js.stripe.com.
11. Add a product to your cart and go to your store's cart page.
12. Make sure we still load scripts from js.stripe.com.
13. Use the Code Snippets plugin to add the following filters:

```php
add_filter( 'wc_stripe_load_scripts_on_product_page_when_prbs_disabled', '__return_false' );
add_filter( 'wc_stripe_load_scripts_on_cart_page_when_prbs_disabled', '__return_false' );
```

14. Open any product details page.
15. Make sure no scripts are loaded from js.stripe.com
16. Add a product to your cart and go to your store's cart page.
17. Make sure no scripts are loaded from js.stripe.com


---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

Commit:
Only load Stripe JS on product page when required

Previously we always loaded the Stripe JS on relevant pages through the
main gateway class, but now we have an additional check there to make
sure that doesn't happen.

Commit:
Only load Stripe JS on the cart page when required

Previously we always loaded the Stripe JS on relevant pages through the
main gateway class, but now we have an additional check there to make
sure this doesn't happen.

Commit:
add changelog entry

diegocurbelo commented on .vscode/launch.json:
```diff
@@ -13,6 +13,12 @@
             }
+        },
+        {
+            "name": "Xdebug local",
```
It has absolutely nothing to do with the PR... but maybe we can name those two something like:
- `Listen for Xdebug (remote)`
- `Listen for Xdebug (local)`

reykjalin replied:
Yeah, this was not supposed to be included 🤦‍♂️ 
I'll remove this. I need to open a separate PR for that.

reykjalin replied:
Removed in 94b2851de80540adc1f34f4ac966747cd7a914eb

diegocurbelo replied:
I meant that my request had nothing to do with the PR... I like the idea of having both Xdebug configs, I did not like the naming though 🤣 . But we can add them in another PR if you prefer..

reykjalin replied:
Haha I see what you mean! I think it should still be part of a separate PR 😛

diegocurbelo reviewed (changes requested):
Code looks good, and it works as expected in both the `product` and `shortcode cart` pages... but it still loads Stripe in the `blocks cart`.

Also, should we create an issue to do this for UPE?

Commit:
Merge branch 'develop' into fix/2065-only-load-stripe-js-on-product-pages-when-required

Commit:
revert launch json config change

reykjalin commented:
> it still loads Stripe in the `blocks cart`.

Yeah, I hadn't thought about it there 🤔 It's less of a problem in the Blocks integration though since the Stripe JS SDK is loaded asynchronously. We should still make removing unnecessary JS an option.

> Also, should we create an issue to do this for UPE?

I think we can just address this here 🤔 

***

@DeanMarkTaylor brought up a point in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2065#issuecomment-960450052 that we've been discussing in p1635879932054100-slack-C01BZUL57SQ regarding how Stripe recommends we load the Stripe JS on all pages for fraud prevention purposes.

Currently we only load the Stripe JS on Product, Cart, and Checkout pages, and I don't think we should change that. However, based on Stripe's own documentation we probably shouldn't be disabling the scripts _by default_.

I think the best path forward would be to introduce a filter that will allow merchants to disable the Stripe JS on product and cart pages when the PRBs are not loaded. We'll still load the Stripe JS by default to follow Stripe's recommendation on loading the script on multiple pages for fraud prevention purposes, but merchants that are willing to accept the risk of _potentially_ increased rates of fraud can disable the script entirely.

Commit:
Add filters to disable scripts on product and cart

These filters will only be applied when the PRBs are disabled on the
product and cart pages. This will give merchants the option to not load
the extra JS when they're not using the PRBs on those pages.

Stripe's JS includes some fraud prevention mechanisms, so disabling the
scripts using these filters does increase the risk of successul fraud on
a merchant's store, but that's something merchants will have to decide
on their own.

reykjalin commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -404,6 +404,32 @@ public function payment_scripts() {
+				true
+			)
+		) {
+			if ( apply_filters( 'wc_stripe_remove_scripts_on_product_page_when_prbs_disabled', false ) ) {
```
These filter names are very much up for discussion. I'm trying to get their function across in the name, but the names are pretty long. Not sure if they can be shortened without losing some of the meaning.

I'd be happy to get some feedback on the naming for both filters.

reykjalin replied:
I changed the filter names to `wc_stripe_load_scripts_on_cart_page_when_prbs_disabled` and `wc_stripe_load_scripts_on_product_page_when_prbs_disabled` to better reflect how they're used in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2110/commits/a152af75ed9d260b36f6ee6637c6457038eb498d

Commit:
Add helper functions to determine when to load js

reykjalin commented on includes/class-wc-stripe-helper.php:
```diff
@@ -583,4 +583,22 @@ public static function should_enqueue_in_current_tab_section( $tab, $section ) {
 	}
+
+	public static function should_load_scripts_on_product_page() {
+		$prb_locations = self::get_settings( null, 'payment_request_button_locations' ) ?? [ 'product', 'cart' ];
```
Automated checks will continue to fail until we merge #2109 since I'm using the null-coalescing operator `??` here, which is exclusive to PHP 7.0.

Commit:
Disable JS for UPE and Blocks when appropriate

Only applies on Product and Cart pages when PRBs are disabled.

reykjalin commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -60,6 +60,14 @@ public function is_active() {
 	 * @return array
 	 */
 	public function get_payment_method_script_handles() {
+		if ( is_product() && ! WC_Stripe_Helper::should_load_scripts_on_product_page() ) {
+			return;
+		}
```
I actually don't think this is necessary. We don't load the blocks on the product page.

Probably better to just change the cart check to see if the `woocommerce/cart` block is present and remove the `is_product()` check entirely 🤔

reykjalin replied:
Based on https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2110#issuecomment-962416878 we're going to always load the scripts in the cart block. It's loaded asynchronously so it shouldn't have a big impact on load times.

reykjalin commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -193,6 +193,14 @@ public function payment_scripts() {
 			return;
 		}
 
+		if ( is_product() && ! WC_Stripe_Helper::should_load_scripts_on_product_page() ) {
+			return;
+		}
+
+		if ( is_cart() && ! WC_Stripe_Helper::should_load_scripts_on_cart_page() ) {
+			return;
+		}
```
I actually don't think this is necessary. Do we run this gateway on the product and cart pages? I feel like we fall back to the base gateway class there. Haven't tested it though. Need to look into it.

reykjalin replied:
Based on testing: the `is_cart()` check is required, but the `is_product()` check is not. I've removed the `is_product()` check in 9bc82b5e8368ed31006eb4ded95555e36df24615.

Commit:
Merge branch 'develop' into fix/2065-only-load-stripe-js-on-product-pages-when-required

diegocurbelo reviewed (changes requested):
After adding the 2 filters... the `product` and `shortcode cart` don't load the Stripe JS, but both pages have a warning at the top (scrolling up after page load), the `blocks cart` does not have the warning but the Stripe JS is still loaded:

<img width="1792" alt="Screen Shot 2021-11-04 at 23 59 04" src="https://user-images.githubusercontent.com/407542/140451070-9bf5140a-f4ce-4164-b3c7-f73489d6013a.png">

Commit:
Merge branch 'develop' into fix/2065-only-load-stripe-js-on-product-pages-when-required

reykjalin commented:
> both pages have a warning at the top (scrolling up after page load)

@diegocurbelo I'm unable to reproduce the issue 🤔 

What versions of WP/WC/Blocks are you using?

> the `blocks cart` does not have the warning but the Stripe JS is still loaded

Ah yeah, I can reproduce this. I'll look into that! Nice catch!

reykjalin commented:
@diegocurbelo It seems maybe we can't detect this for the blocks integration?

When I'm on the page with the cart block (designated as the **Cart** page) both `is_cart()` and `has_block( 'woocommerce/cart' )` return `false` 🤨 .

<img width="929" alt="image" src="https://user-images.githubusercontent.com/13835680/140602961-2459d478-a389-49ca-b4aa-18ecb116ac81.png">

This is definitely less of a problem with the Blocks integration because the Stripe JS is loaded asynchronously via the Stripe SDK npm package using `loadStripe()`, but it seems like we can't offer this filter in the Blocks integration.

If we only rely on the settings (the result from `WC_Stripe_Helper::should_load_scripts_on_cart_page()`) we'd end up removing the script from the Checkout page too, and that's not an option.

All in all it looks like we may have to just not apply the filter for the blocks integration? What do you think?

diegocurbelo commented:
> All in all it looks like we may have to just not apply the filter for the blocks integration? What do you think?

I agree... The JS loads asynchronously in that case

Commit:
return empty list of script handles for blocks support

Commit:
Merge branch 'develop' into fix/2065-only-load-stripe-js-on-product-pages-when-required

Commit:
Remove filter for unloading Stripe JS with blocks

The function used to retrieve the script handles is run at a point where
the block hasn't been loaded, so we can't detect whether we're on a cart
or product page. As such we can't make an informed decisions on whether
or not to load the script.

The script is loaded asynchronously in the blocks checkout though, so
this shouldn't be degradation in performance as it relates to the issue
reported in #2065.

Commit:
Remove redundant product page check for UPE filter

The UPE payment_scripts() function is not called on product pages, and
as such it is not necessary to check for that condition when unloading
the Stripe JS.

diegocurbelo reviewed (approved):
Code looks good... and tests well 🚢 🚢
